### PR TITLE
feat (web/server) 360 degrees Web panoramas

### PR DESCRIPTION
**Pull request by dmitry-brazhenko**
In this PR I aded
1. Support for 360 Panoramas as it was discussed here: https://github.com/immich-app/immich/discussions/1630
2. Minor technical changes and improvements

**How it was implemented**
1. Once an images is uploaded, in [metadata-extraction.processor.ts](https://github.com/immich-app/immich/pull/3098/files#diff-ed7be838d2a88fd5de449b030f1fc0952987eba3b56897b7cb68a8b1442edbd0) it checks aspectRatio. If it is aspectRatio >= 2, we assume that it is a panorama. This can be potentialy improved, but looks like a common way to determine if image is a panorama or not
2. If it is a panorama, we add a prtoperty isPanorama: true
3. We use library @egjs/svelte-view360 to show the panoram
4. In the library we use a special symbol to display if a thumb is a panorama. 

**Minor technical changes and improvements:**
1. restart:always for all docker containers in development docker-compose
2. Fixed links in database-migrations.md file

**Potential further improvements:**
1. Probably we can switch library and use smth else. But this liibrary works OK both on desktop and mobile-web
2. Support panoramas in Flutter mobile app 
3. There 360 videos. Later we can support them as well

Here is thumb that shows a panorama:

Panorama after is opened:

### CODE REVIEW

**Comment by dmitry-brazhenko** [comment]
Probably I need to add some tests, but not sure. Need some advise. 
Also I am not 100% sure that I properly fixed open-api-doc. Please help me with this stuff. 
Thanks

File: web/src/lib/components/assets/thumbnail/360arrow.svg
Comment (jrasm91) [inline web/src/lib/components/assets/thumbnail/360arrow.svg:1]: We can probably use an existing icon from the icon library instead of loading one of these

https://pictogrammers.com/library/mdi/

image

eg.
```typescript
import Rotate360Icon from 'svelte-material-icons/Rotate360.svelte'
```

**Review by jrasm91 — COMMENTED** [review]
You can try running `make api` to regenerate the api. You can also run the `npm run check:all` scripts in the web/server folders to make sure those all pass.

File: mobile/openapi/pubspec.yaml [hunk 1]
```diff
@@ -10,7 +10,7 @@ environment:
   sdk: '>=2.12.0 <3.0.0'
 dependencies:
   http: '>=0.13.0 <0.14.0'
-  intl: '^0.18.0'
+  intl: '^0.17.0'
```
Comment (alextran1502) [inline mobile/openapi/pubspec.yaml:13]: Is this change intentional?

Comment (dmitry-brazhenko) [reply mobile/openapi/pubspec.yaml:13]: It is not. It was made automatically once I ran

File: web/src/lib/components/assets/thumbnail/360arrow.svg
Comment (dmitry-brazhenko) [reply web/src/lib/components/assets/thumbnail/360arrow.svg:1]: Yes, I ran npm run check:all and api generator. 

Thanks for sharing these icons!!
I reused the first one.

**Comment by dmitry-brazhenko** [comment]
For further steps for improvement I see to add a "panorama" filter here:

File: web/src/lib/components/assets/thumbnail/image-thumbnail.svelte [hunk 2]
```diff
@@ -21,7 +21,7 @@
 {#if displayPanoramaArrow}
   <div class="absolute right-0 top-0 text-white text-xs font-medium flex gap-1 place-items-center z-20">
     <span class="pt-2 pr-2">
-      <img src={PanoramaArrow} alt="" width="24" height="24" />
+      <svelte:component this={Rotate360Icon} size="24" />
```
Comment (jrasm91) [inline web/src/lib/components/assets/thumbnail/image-thumbnail.svelte:24]: `<Rotate360Icon size="24"/>`

Comment (dmitry-brazhenko) [reply web/src/lib/components/assets/thumbnail/image-thumbnail.svelte]: Thanks, fix ed

**Comment by dmitry-brazhenko** [comment]
Hi!
Do you have any ideas what can be improved in this PR? Can we merge it? :)

**Comment by alextran1502** [comment]
@dmitry-brazhenko hello, I haven't gotten a chance to review the PR yet. Once I do and if there is nothing else to address, it will be merged. Thank you for the contribution

File: web/src/lib/components/asset-viewer/panorama-viewer.css
Comment (jrasm91) [inline web/src/lib/components/asset-viewer/panorama-viewer.css:1]: Is it possible to include this css in the asset-viewer.svelte component directly via a `<style> ... </style>` tag at the end of the file instead?

File: web/src/lib/components/asset-viewer/panorama-viewer.svelte [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  const loadAssetData = async () => {
+    try {
+      const { data } = await api.assetApi.serveFile(
+        { id: asset.id, isThumb: false, isWeb: false, key: publicSharedKey },
+        { responseType: 'blob' },
+      );
+
+      if (data instanceof Blob) {
+        assetData = URL.createObjectURL(data);
```
Comment (jrasm91) [inline web/src/lib/components/asset-viewer/panorama-viewer.svelte:22]: assetData is a data url, right? Can we name this variable something like `fileUrl`, `dataUrl` instead?

File: web/src/lib/components/assets/thumbnail/image-thumbnail.svelte [hunk 4]
```diff
@@ -12,10 +13,19 @@
...
 </script>
 
+{#if displayPanoramaArrow}
+  <div class="absolute right-0 top-0 text-white text-xs font-medium flex gap-1 place-items-center z-20">
+    <span class="pt-2 pr-2">
+      <Rotate360Icon size="24" />
+    </span>
+  </div>
+{/if}
+
```
Comment (jrasm91) [inline web/src/lib/components/assets/thumbnail/image-thumbnail.svelte:28]: I don't think this is the correct place for this icon. We have other icons on thumbnails to indicate if it is archived, favorited, etc. Those live in `web/src/lib/components/assets/thumbnail/thumbnail.svelte`. I believe that would be the appropriate place for this as well. It also removes the need for `displayPanoramaArrow` in this component.

File: server/src/microservices/processors/metadata-extraction.processor.ts [hunk 5]
```diff
@@ -200,8 +200,19 @@ export class MetadataExtractionProcessor {
...
+      const aspectRatio = newExif.exifImageWidth / newExif.exifImageHeight;
+      isPanorama = aspectRatio >= 2; // This is a standard way to determine if an image is a panorama
+    }
+
     await this.exifRepository.upsert(newExif, { conflictPaths: ['assetId'] });
-    await this.assetRepository.save({ id: asset.id, fileCreatedAt: fileCreatedAt || undefined });
+    await this.assetRepository.save({
+      id: asset.id,
+      fileCreatedAt: fileCreatedAt || undefined,
+      isPanorama: isPanorama,
```
[the remark this answers is not in the document]
Comment (dmitry-brazhenko) [reply server/src/microservices/processors/metadata-extraction.processor.ts:214]: I would suggest to keep named parameter to reduce the number of potential mistakes

File: web/src/lib/components/asset-viewer/panorama-viewer.css
Comment (dmitry-brazhenko) [reply web/src/lib/components/asset-viewer/panorama-viewer.css:1]: I tried to do that, but it did not work. Here is a lib example: https://github.com/naver/egjs-view360/blob/master/packages/svelte-view360/demo/App.svelte

I overrided original css file. I asked a question [here](https://github.com/naver/egjs-view360/issues/424), but the suggestion did not work. 
I am not very good in frontend technologies. If there are any ways to enable, please let me know, I will fix it :)

File: web/src/lib/components/asset-viewer/panorama-viewer.svelte (hunk 3 above)
Comment (dmitry-brazhenko) [reply web/src/lib/components/asset-viewer/panorama-viewer.svelte:22]: Sure, renamed it, thanks :) 
to be honest, I reused naming from photo-viewer component: https://github.com/immich-app/immich/blob/8a3ab5be3e2628a5c777876eab3dc23ac519f2e6/web/src/lib/components/asset-viewer/photo-viewer.svelte#L66

File: web/src/lib/components/assets/thumbnail/image-thumbnail.svelte (hunk 4 above)
Comment (dmitry-brazhenko) [reply web/src/lib/components/assets/thumbnail/image-thumbnail.svelte]: I moved it to thumbnail.svelte because panorama icon can be potentially displayed for videos as well. 

But as far as I see here video-related icons are rendered in video-thumbnail component
https://github.com/immich-app/immich/blob/8a3ab5be3e2628a5c777876eab3dc23ac519f2e6/web/src/lib/components/assets/thumbnail/video-thumbnail.svelte

File: server/src/microservices/processors/metadata-extraction.processor.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jrasm91) [reply server/src/microservices/processors/metadata-extraction.processor.ts:214]: How does this last to more potential mistakes? It is just a shorthand syntax for the same thing.

File: web/src/lib/components/asset-viewer/panorama-viewer.css
Comment (jrasm91) [reply web/src/lib/components/asset-viewer/panorama-viewer.css:1]: I think it can work, but you might need to add something special to make the style not scoped to the component.

File: server/src/microservices/processors/metadata-extraction.processor.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmitry-brazhenko) [reply server/src/microservices/processors/metadata-extraction.processor.ts]: Oh, I did not know about this syntax. Thanks for sharing!

File: web/src/lib/components/asset-viewer/panorama-viewer.svelte (hunk 3 above)
Comment (jrasm91) [reply web/src/lib/components/asset-viewer/panorama-viewer.svelte]: Ah makes sense. That was before my time 😛.

File: web/src/lib/components/assets/thumbnail/image-thumbnail.svelte (hunk 4 above)
Comment (jrasm91) [reply web/src/lib/components/assets/thumbnail/image-thumbnail.svelte]: Yeah, we can cross that bridge when we get there. I think this is the right place for now.

File: server/openapi-generator/templates/web/apiInner.mustache [hunk 6]
```diff
@@ -173,9 +173,8 @@ export const {{classname}}AxiosParamCreator = function (configuration?: Configur
             {{^isArray}}
             if ({{paramName}} !== undefined) { {{^multipartFormData}}
                 localVarFormParams.set('{{baseName}}', {{paramName}} as any);{{/multipartFormData}}{{#multipartFormData}}{{#isPrimitiveType}}
-                localVarFormParams.append('{{baseName}}', {{paramName}} as any);{{/isPrimitiveType}}{{^isPrimitiveType}}{{#isEnum}}
-                localVarFormParams.append('{{baseName}}', {{paramName}} as any);{{/isEnum}}{{^isEnum}}
-                localVarFormParams.append('{{baseName}}', new Blob([JSON.stringify({{paramName}})], { type: "application/json", }));{{/isEnum}}{{/isPrimitiveType}}{{/multipartFormData}}
+                localVarFormParams.append('{{baseName}}', {{paramName}} as any);{{/isPrimitiveType}}{{^isPrimitiveType}}
+                localVarFormParams.append('{{baseName}}', new Blob([JSON.stringify({{paramName}})], { type: "application/json", }));{{/isPrimitiveType}}{{/multipartFormData}}
```
Comment (jrasm91) [inline server/openapi-generator/templates/web/apiInner.mustache:177]: Need to double check where this change is coming from

**Review by jrasm91 — APPROVED** [review]
Need to double check the API generated code, but other than that LGTM.

**Comment by alextran1502** [comment]
I think there is some issue with the recent rebase/merge, can you only include the changes to the panorama view?

**Comment by dmitry-brazhenko** [comment]
After I ran "make api" - it resulted to multiple file changes

**Comment by alextran1502** [comment]
These changes somehow reverting some of the changes that we made. How about you only include the changes in the `server` and `mobile` and leave the API generation for me?

**Comment by dmitry-brazhenko** [comment]
Yes, I will do that

**Comment by dmitry-brazhenko** [comment]
@alextran1502 I reverted my attempts to merge main branch to my branch. 
Could you help me with merging please? 
Thanks

File: server/openapi-generator/templates/web/apiInner.mustache
Comment (alextran1502) [inline server/openapi-generator/templates/web/apiInner.mustache:1]: Please remove this file from your PR

File: docker/docker-compose.dev.yml
Comment (alextran1502) [inline docker/docker-compose.dev.yml:1]: Please remove this file from your PR

File: docs/docs/developer/database-migrations.md
Comment (alextran1502) [inline docs/docs/developer/database-migrations.md:1]: Please remove this file from your PR

Comment (dmitry-brazhenko) [reply docs/docs/developer/database-migrations.md:1]: Why remove it? I fixed the path? 
https://github.com/immich-app/immich/tree/main/server/src/infra/entities

File: docker/docker-compose.dev.yml
Comment (dmitry-brazhenko) [reply docker/docker-compose.dev.yml:1]: Why not to keep it? 
Once I started running Immich, it was failing as containers just stopped. Adding this restart;always helped me to get started

**Comment by dmitry-brazhenko** [comment]
@alextran1502 would you mind if Ierge docker compose improvement and MD fix in a separate PR?

**Comment by uhthomas** [comment]
Fixes #2610.

**Comment by brighteyed** [comment]
>  If it is aspectRatio >= 2, we assume that it is a panorama

I think 360 panoramas should be detected more reliably in this PR. Currently, the code assumes that all images with an aspect ratio >= 2 are 360 panoramas. 

But, for example:
 - Screenshots with landscape orientation should not be detected as 360 panoramas
 - Non-360 panoramas (for example, with cylindrical projection) should not be detected as 360 panoramas

I believe we should use the EXIF tag `ProjectionType` to detect 360 panoramas in addition to the aspect ratio check

**Comment by CodeWithMa** [comment]
> > If it is aspectRatio >= 2, we assume that it is a panorama
> 
> I think 360 panoramas should be detected more reliably in this PR. Currently, the code assumes that all images with an aspect ratio >= 2 are 360 panoramas.
> 
> But, for example:
> 
>     * Screenshots with landscape orientation should not be detected as 360 panoramas
> 
>     * Non-360 panoramas (for example, with cylindrical projection) should not be detected as 360 panoramas
> 
> 
> I believe we should use the EXIF tag `ProjectionType` to detect 360 panoramas in addition to the aspect ratio check

I agree. My google pixel photosphere images have the following projection type:

```
Projection Type                 : equirectangular
```

I also have photoshere images that are not complete 360 degree images so the aspect ratio is not 2:1.

**Comment by dmitry-brazhenko** [comment]
@CodeWithMa @brighteyed  you're right. I fixed panorama detection. Now it consumes Exif and marks images by "panorama" only if there is a necessary exif tag. 

@alextran1502 Could you help me with merging this PR? 

Thanks

**Comment by alextran1502** [comment]
Can you create another PR to consolidate the changes on the web and the server? Please put migrations into a single file. Once you've done that please ping me and I can help generating the API

File: server/src/domain/asset/response-dto/asset-response.dto.ts [hunk 7]
```diff
@@ -32,6 +32,7 @@ export class AssetResponseDto {
   people?: PersonResponseDto[];
   /**base64 encoded sha1 hash */
   checksum!: string;
+  isPanorama!: boolean;
```
Comment (brighteyed) [inline server/src/domain/asset/response-dto/asset-response.dto.ts:35]: I'd prefer `projectionType` field with the `enum` type

File: server/src/domain/asset/response-dto/asset-response.dto.ts [hunk 8]
```diff
@@ -58,6 +59,7 @@ export function mapAsset(entity: AssetEntity): AssetResponseDto {
     tags: entity.tags?.map(mapTag),
     people: entity.faces?.map(mapFace),
     checksum: entity.checksum.toString('base64'),
+    isPanorama: entity.isPanorama,
```
Comment (brighteyed) [inline server/src/domain/asset/response-dto/asset-response.dto.ts:62]: I'd prefer `projectionType` field with the `enum` type

File: server/src/infra/entities/exif.entity.ts [hunk 9]
```diff
@@ -43,6 +43,9 @@ export class ExifEntity {
   @Column({ type: 'float', nullable: true })
   longitude!: number | null;
 
+  @Column({ type: 'varchar', nullable: true })
+  projectionType!: string | null;
```
Comment (brighteyed) [inline server/src/infra/entities/exif.entity.ts:47]: Can we use `enum` here?

File: server/package.json [hunk 10]
```diff
@@ -59,11 +59,12 @@
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.0",
     "cookie-parser": "^1.4.6",
-    "exiftool-vendored": "^19.0.0",
+    "exiftool-vendored": "^22.0.0",
```
Comment (brighteyed) [inline server/package.json:62]: Do we need to update `exiftool-vendored` in this PR?

**Review by brighteyed — COMMENTED** [review]
If we transform `isPanorama` fields into `projectionType` it will be easier in the future to add additional types of projections (most likely `cylindrical`)

File: server/package.json (hunk 10 above)
Comment (dmitry-brazhenko) [reply server/package.json:62]: Yes, we need to update it as ProjectionType does not exist in previous version

**Comment by dmitry-brazhenko** [comment]
Will go on in this PR: https://github.com/immich-app/immich/pull/3412. 

Implemented Projectiontype as enum
